File: src/web-paddle-control/src/mqttClient.js
import mqtt from 'mqtt';

class MQTTClient {
  constructor() {
    this.client = null;
    this.onConnectCallback = null;
  }

  connect(onConnectCallback) {
    const broker = process.env.REACT_APP_MQTT_BROKER;
    const port = process.env.REACT_APP_MQTT_PORT;
    const clientId = process.env.REACT_APP_MQTT_CLIENT_ID;

    this.client = mqtt.connect(`ws://${broker}:${port}`, { clientId });
    this.onConnectCallback = onConnectCallback;

    this.client.on('connect', () => {
      console.log('Connected to MQTT broker');
      if (this.onConnectCallback) {
        this.onConnectCallback();
      }
    });

    this.client.on('disconnect', () => {
      console.log('Disconnected from MQTT broker');
    });

    this.client.on('error', (err) => {
      console.error('Connection error: ', err);
      this.client.end();
    });

    this.client.on('reconnect', () => {
      console.log('Reconnecting...');
    });
  }

  isConnected() {
    return this.client && this.client.connected;
  }

  publish(topic, message) {
    if (this.isConnected()) {
      this.client.publish(topic, message);
    }
  }
}

export default MQTTClient;

File: src/keyboard-paddle-control/src/mqtt_client.py
import paho.mqtt.client as mqtt

class MQTTClient:
    def __init__(self, mqtt_broker, mqtt_port, mqtt_client_id):
        self.mqtt_broker = mqtt_broker
        self.mqtt_port = mqtt_port
        self.mqtt_client_id = mqtt_client_id
        self.connected = False
        self.client = mqtt.Client(self.mqtt_client_id)
        self.client.on_connect = self.on_connect
        self.client.on_disconnect = self.on_disconnect
        # self.client.on_publish = self.on_publish

    def on_connect(self, client, userdata, flags, rc):
        # print(f"Connected with result code {rc}")
        self.connected = True

    # def on_publish(self, client, userdata, mid):
    #     print(f"Message {mid} published.")

    def on_disconnect(self, client, userdata, rc):
        # print(f"Disconnected with result code {rc}")
        self.connected = False

    def connect(self):
        self.client.connect(self.mqtt_broker, self.mqtt_port, 60)
        self.client.loop_start()

    def publish(self, topic, message):
        if self.is_connected():
            self.client.publish(topic, message)
        else:
            print("MQTT client is not connected.")

    def disconnect(self):
        self.client.loop_stop()
        self.client.disconnect()

    def is_connected(self):
        return self.connected
